### ISSUE TITLE
Crawler4j not completely finishing after redirect

### ISSUE THREAD

**Issue by brbog** [rzo1/crawler4j#95]
Context:
It's only happening during my tests with some Joomla site that requires form logon (nothing personal, I can share the username and password until the end of August). After the redirect (via http status 303) crawling stops because of my configuration that says not to follow links in the login page (which is ok I think). I'm still investigating further how the authentication is supposed to work.

Problem(s):
- I suspect the sources for the frontiers are not uploaded to Maven Central Repository, so I can 't debug inside them.
- After seeing that there are no new urls to crawl, `frontier.isFinished()` is called inside the `WebCrawler`-class. The application ends, but there are still some threads endlessly continuing (see screenshot). This issue doesn't happen for my other tests without authentication.

It's not a priority as I suspect that once my authentication is working things may work out fine, but created the issue in case this is of intrest. The HSQLDB frontier is being used.

![crawler4j-not-finishing](https://user-images.githubusercontent.com/100319987/178136128-39c5f93d-0c90-4652-91fc-379a640ec101.png)

**Comment by rzo1** [comment]
Sources for the frontier impls should be available on Maven Central (they are a sync requirement), Perhaps they are not attached, if the aggregations POM is used?

- https://repo1.maven.org/maven2/de/hs-heilbronn/mi/crawler4j-frontier-sleepycat/4.9.1/
- https://repo1.maven.org/maven2/de/hs-heilbronn/mi/crawler4j-frontier-urlfrontier/4.9.1/
- https://repo1.maven.org/maven2/de/hs-heilbronn/mi/crawler4j-frontier-hsqldb/4.9.1/

**Comment by brbog** [comment]
Attached the content of the DB after "completion" for completeness.
![crawler4j-not-finishing-db-content](https://user-images.githubusercontent.com/100319987/178136407-f1765836-a8ff-4854-a739-87263d3dafce.png)

**Comment by brbog** [comment]
> Sources for the frontier impls should be available on Maven Central (they are a sync requirement), Perhaps they are not attached, if the aggregations POM is used?
> 
>     * https://repo1.maven.org/maven2/de/hs-heilbronn/mi/crawler4j-frontier-sleepycat/4.9.1/
> 
>     * https://repo1.maven.org/maven2/de/hs-heilbronn/mi/crawler4j-frontier-urlfrontier/4.9.1/
> 
>     * https://repo1.maven.org/maven2/de/hs-heilbronn/mi/crawler4j-frontier-hsqldb/4.9.1/

You 're right, something weird was going on where I had WebCrawler.java and WebCrawler.class as tabs inside Eclipse, now I can get to the code. One problem down :-). Not immediately clear why these threads don't stop though.

**Comment by rzo1** [comment]
I guess, that we need to profile (VisualVM, YourKit, etc) in order to get a clear view. Probably something is blocking, but didn't look ed at the code yet :)

**Comment by brbog** [comment]
Sorry, my bad... I had copied the PageFetcher to a new class and extended it to override the url resolving on a status 303.
Looks like this: `public class BbPageFetcher extends PageFetcher {`.
The problem is that my quick and dirty copy didn't call super in the shutdown() method:
```Java
public synchronized void shutDown() {
        if (connectionMonitorThread != null) {
            connectionManager.close();
            connectionMonitorThread.shutdown();
        }
    }
```
Turns out that PageFetcher is very easy to extend because everything is mostly protected, so deleting everything but the fetchPage()-method fixed it.
